Upsun Quick start guide (#3554)
platformsh/platformsh-docs: sites/friday/src/get-started/here/make-changes.md

---
title: Make changes to your project
sidebarTitle: Revisions
weight: 50
description: How to work on a daily basis with an {{% vendor/name %}} project?
---

Now that your project is deployed, you can start making changes to it.
For example, you might want to fix a bug or add a new feature.
It's this component of Upsun - preview environments - that you'll come to recognize as a key feature.

> In your project, the `main` branch (or `master` if you change it during project creation) always represents the production environment.

Reviewer: "In your project, the project's `default_branch` (eg `main`, `master`, or whatever you entered during project creation)..."

Other branches are for developing new features, fixing bugs, or updating the infrastructure.

To make changes to your project, follow these steps:

1. Create a new environment (a Git branch) to make changes without impacting production:
{{< codetabs >}}
+++
title={{% vendor/name %}} Git repository
+++
   ```bash {location="Terminal"}

>    {{% vendor/cli %}} branch feat-a

Reviewer: given we're trying to juggle integrations and non-integrations at this point, is there any advantage to instructing the user to use the our cli to create the branch over just using git? By using our CLI we need to include additional steps for the integration path.

   ```
   This command creates a new local `feat-a` Git branch based on the `main` Git branch
   and activates a related environment on {{< vendor/name >}}.
   The new environment inherits the data (service data and assets) of its parent environment (the production environment here).

>    {{< note title="Warnings" >}}

Reviewer: If we change to having the user use git for branching instead of our cli, then we can move the warning out of this tab, and below it's conclusion since the warning is applicable to both scenarios.

Reviewer: no, unfortunately i think it's even *more* confusing now. 

My suggestion is to remove the codetabs for `Upsun Git repository` and `Using third party provider`, remove `upsun branch feat-a` and replace it with `git branch checkout -b feat-a`.

This way, it doesn't matter if they're using the Upsun repository or a 3rd party, the command is the same. And given at this stage nothing has changed, there really isn't a need to push yet.

We can move all of this down to step 4 when it's time to push. At that point, we can split between the two scenarios.

>    **First**, you will be charged by [activated environments](/environments/deactivate-environment.md#reactivate-an-environment).
>    So, if you would like to create an inactive environment, waiting for you to be ready to test, you could use the following command:
>    ```bash {location="Terminal"}
>    git branch checkout -b feat-a && git push -u upsun feat-a
>    ```

Reviewer 2: ```suggestion
   **First**, you are charged by [activated environments](/environments/deactivate-environment.md#reactivate-an-environment).
    So, to create an inactive environment waiting for you to be ready to test, run the following command:
   ```bash {location="Terminal"}
   git branch checkout -b feat-a && git push -u upsun feat-a
   ```
```

>    **Secondly**, if you have set up an integration and have enabled `prune_branches` (true by default), then `upsun branch` command will fail at the end of the deployment because the integration will sync with the source and delete the new branch since it doesn’t exist with the canonical source.

Reviewer: shouldn't this warning be on the "using third party provider" tab? If the user has a local repo and isn't using an integration, they wont have this issue.

   At the end, you will see:
   ```bash
   fatal: couldn't find remote ref feat-a
   ```
   {{< /note >}}
<--->
+++
title=Using third party provider
+++
   {{% vendor/name %}} provides a feature called [source integration](integrations/source.html) that allows your {{% vendor/name %}} project to be fully integrated with your external repository.
   This enables you, as a developer, to use a normal Git workflow (`git add . && git commit -m "message" && git push`) to deploy your environment—with no need to connect to the {{% vendor/name %}} Console.

   Assuming you already activate source integration feature on your project, you need to use the following to create a new branch:
   ```bash {location="Terminal"}
   git checkout -b feat-a && git push -u origin feat-a
   ```
   This command creates a new local `feat-a` Git branch based on the `main` Git branch
   and create related environment on {{< vendor/name >}}.
   The new environment is inactive by default, and when you would activate it, it will inherit the data (service data and assets) of its parent environment (the production environment here).

>    {{< note >}}

Reviewer: shouldnt this be an included step instead of note? Later at step 4 we've _assumed_ they've activated it, but this note makes it appear as if it is an optional item.

Author: What do you mean? 
is it ok if i just add a note block on the Deploy your changes, telling the reader that if the environment is inactive, he needs to activate it ?
https://docs.upsun.com.fix-generic-gstarted-7jc6g4y-ucq44jg6ofare.eu-5.platformsh.site/get-started/here/make-changes.html#4-deploy-your-changes

Reviewer: see https://docs.upsun.com.fix-generic-gstarted-7jc6g4y-ucq44jg6ofare.eu-5.platformsh.site/get-started/here/make-changes.html#4-deploy-your-changes

   When your local feature is ready, please see how to [activate your environment](/environments/deactivate-environment.md#reactivate-an-environment)
   {{< /note >}}

{{< /codetabs >}}

2. Make changes to your project.

   Depending on the stack you're using, change something within your source code.

3. Commit your changes:

   ```bash {location="Terminal"}
   git add index.js
   git commit -m "Update source code"
   ```

> 4. Deploy your changes to the `feat-a` environment:

Reviewer: If they are using an integration, since we did not instruct them to activate the environment after pushing to origin, the environment may not be active

Author: same as previous comment: i add a new note block telling the reader to activate it (if not)

Reviewer: see https://github.com/platformsh/platformsh-docs/pull/3554#issuecomment-1949231306

   ```bash {location="Terminal"}
   {{% vendor/cli %}} push
   ```

   Note that each environment has its own domain name.
   To open the url of your new environment, run the following command:

   ```bash {location="Terminal"}

>    {{% vendor/cli %}} environment:url --primary

Reviewer: If they are using an integration, since we did not instruct them to activate the environment after pushing to origin, this environment may not be active and therefore the user will receive an error when they attempt to retrieve the environment URL

Author: same as previous comment: i add a new note block telling the reader to activate it (if not)

Reviewer: see https://github.com/platformsh/platformsh-docs/pull/3554#issuecomment-1949231306

   ```

5. Iterate by changing the code, committing, and deploying.
   When satisfied with your changes, merge them to the main branch,
   and remove the feature branch:

   ```bash {location="Terminal"}
   {{% vendor/cli %}} merge
     Are you sure you want to merge feat-a into its parent, main? [Y/n] y
   {{% vendor/cli %}} checkout main
   git pull {{% vendor/cli %}} main
   {{% vendor/cli %}} environment:delete feat-a
   git fetch --prune
   ```

   {{< note >}}
   Deploying to production was fast because the image built for the `feat-a` environment was reused for your `main` environment.
   {{< /note >}}

>    For a long running branch, to keep the code up-to-date with the main branch, use `git merge main` or `git rebase main`.

Reviewer: this seems out of place? I assume what we're suggesting is incorporating changes from `main` back into dev branches, but in step 5 we had them switch back to the `main` branch so attempting `merge main` or `rebase main` when you have `main` checked out doesn't do anything.

Author: i make it a real 6. step
https://docs.upsun.com.fix-generic-gstarted-7jc6g4y-ucq44jg6ofare.eu-5.platformsh.site/get-started/here/make-changes.html#4-deploy-your-changes
is that ok?

Reviewer: see https://github.com/platformsh/platformsh-docs/pull/3554#issuecomment-1949231306

---
title: Make changes to your project
sidebarTitle: Revisions
weight: 50
description: How to work on a daily basis with an {{% vendor/name %}} project?
---

Upsun allows you to make changes to your project and test them on a preview environment before introducing them to Production.

In your project, the default branch (e.g. `main`, `master`, whichever chosen during project creation) always represents the production environment.
Other branches are for developing new features, fixing bugs, or updating the infrastructure.

To make changes to your project, follow these steps:

## 1. Create a new environment
Create a new environment (a Git branch) to make changes without impacting production:
{{< codetabs >}}
+++
title={{% vendor/name %}} Git repository
+++

Run the following command:
   ```bash {location="Terminal"}
   {{% vendor/cli %}} branch feat-a
   ```
   This creates a new local `feat-a` Git branch based on the `main` Git branch
   and activates a related environment on {{< vendor/name >}}.
   The new environment inherits the data (service data and assets) of its parent environment (the production environment here).
<--->
+++

> title=Using third party provider

Author: We didn't remove 3rd party integration so do i still need to move it to its own section?

+++
   {{% vendor/name %}} provides a feature called [source integration](integrations/source.html) that allows your {{% vendor/name %}} project to be fully integrated with your external repository.
   This enables you, as a developer, to use a normal Git workflow (`git add . && git commit -m "message" && git push`) to deploy your environment—with no need to connect to the {{% vendor/name %}} Console.

   Assuming you've already activated the source integration feature on your project, create a new branch:
   ```bash {location="Terminal"}
   git checkout -b feat-a && git push -u origin feat-a
   ```
   This creates a new local `feat-a` Git branch based on the `main` Git branch
   and a related environment on {{< vendor/name >}}.
   The new environment is inactive by default. Once you activate it, it will inherit the data (service data and assets) of its parent environment (the Production environment here).

   {{< note >}}
   When your local feature is ready, [activate your environment](/administration/cli/reference.html#environmentactivate).
   {{< /note >}}

{{< /codetabs >}}

{{< note title="Warnings" theme="warning">}}

>    **First**, you are only charged for your [activated environments](/administration/cli/reference.html#environmentactivate).

Reviewer: Re-reading this again, we should probably link to the definition of an "active environment" BUT the definition we have currently seems incomplete. 
> An environment that’s deployed. 

But what does that mean? 
```suggestion
   **First**, you are charged for all [activated environments](/administration/cli/reference.html#environmentactivate).
```

   So, to create an inactive environment waiting for you to be ready to test, run the following command:
   ```bash {location="Terminal"}
   git branch checkout -b feat-a && git push -u upsun feat-a
   ```

>    **Secondly**, if you have set up a [source integration](/integrations/source/_index.md), make sure you disable the `prune_branches` option (`true` by default). Otherwise, the `upsun branch` command fails at the end of the deployment.

Reviewer: if we keep the split path between `Upsun Git repository` and `Using third party provider`, we should move this second warning to the bottom of the `Using third party provider`section just after we have them activate their environment.

Author: i moved it inside the third party provider tab but now, i feel uncomfortable with making this warning hidden at first look (need to select the third party provider to read it). 
https://docs.upsun.com.fix-generic-gstarted-7jc6g4y-ucq44jg6ofare.eu-5.platformsh.site/get-started/here/make-changes.html#1-create-a-new-environment

   and activates a related environment on {{< vendor/name >}}.
   The new environment inherits the data (service data and assets) of its parent environment (the production environment here).
<--->
+++
title=Using third party provider
+++
   {{% vendor/name %}} provides a feature called [source integration](integrations/source.html) that allows your {{% vendor/name %}} project to be fully integrated with your external repository.
   This enables you, as a developer, to use a normal Git workflow (`git add . && git commit -m "message" && git push`) to deploy your environment—with no need to connect to the {{% vendor/name %}} Console.

   Assuming you've already activated the source integration feature on your project, create a new branch:
   ```bash {location="Terminal"}
   git checkout -b feat-a && git push -u origin feat-a
   ```
   This creates a new local `feat-a` Git branch based on the `main` Git branch
   and a related environment on {{< vendor/name >}}.
   The new environment is inactive by default. When you would activate it, it will inherit the data (service data and assets) of its parent environment (the Production environment here).

   {{< note >}}

>    When your local feature is ready, [activate your environment](/environments/deactivate-environment.md#reactivate-an-environment).

Reviewer: ```suggestion
   When your local feature is ready, [activate your environment](/administration/cli/reference.html#environmentactivate).
```

   {{< /note >}}

{{< /codetabs >}}

{{< note title="Warnings" theme="warning">}}

>    **First**, you are only charged for your [activated environments](/environments/deactivate-environment.md#reactivate-an-environment).

Reviewer: ```suggestion
   **First**, you are only charged for your [activated environments](/administration/cli/reference.html#environmentactivate).
```

   ```bash {location="Terminal"}
   fatal: couldn't find remote ref feat-a
   ```
{{< /note >}}

## 2. Make changes to your project.

Depending on the stack you're using, change something within your source code.

## 3. Commit your changes

```bash {location="Terminal"}
git add index.js
git commit -m "Update source code"
```

## 4. Deploy your changes

Deploy your changes to the `feat-a` environment:

```bash {location="Terminal"}
{{% vendor/cli %}} push
```

Note that each environment has its own domain name.
To open the URL of your new environment, run the following command:

```bash {location="Terminal"}
{{% vendor/cli %}} environment:url --primary
```
{{< note title="Warnings" theme="warning">}}
If your environment is inactive, you need to [activate it](/environments/deactivate-environment.md#reactivate-an-environment) first, using the following:
```bash {location="Terminal"}
{{% vendor/cli %}} environment:activate
```
{{< /note >}}

## 5. Iterate
Iterate by changing the code, committing, and deploying.
When satisfied with your changes, merge them to the main branch,
and remove the feature branch:

```bash {location="Terminal"}

> {{% vendor/cli %}} merge

Reviewer: we're going to need to split this into codetabs again for non-integrated vs integrated since the flow can be different, and will break if the default/production branch is protected in the source control management system (ie GitHub)

Reviewer: i'm confused why this was marked resolved since it is still relevant. We're still splitting the path currently between `Upsun Git repository` and `Using third party provider` in step 1, so there's still a significant probability that attempting to perform a merge via our cli tool at step 5 will fail when using a 3rd party integration.

Author: ok, i try to split any git/upsun snippet that could lead to errors/confusion using codetabs.

 Are you sure you want to merge feat-a into its parent, main? [Y/n] y
{{% vendor/cli %}} checkout main
git pull {{% vendor/cli %}} main
{{% vendor/cli %}} environment:delete feat-a
git fetch --prune
```

{{< note >}}
Deploying to production was fast because the image built for the `feat-a` environment was reused for your `main` environment.
{{< /note >}}

## 6. Keep your branch up to date
For a long running dev branch, to keep the code up-to-date with the `main` branch, use `git merge main` or `git rebase main`.

> You can also keep the data in sync with the parent environment by using `{{% vendor/cli %}} env:sync`.

Reviewer: ```suggestion
You can also keep the data in sync with the parent environment by using [`{{% vendor/cli %}} env:sync data`](/administration/cli/reference.html#environmentsynchronize).
```
